installer/installer.go

```diff
@@ -0,0 +1,821 @@
+
+	savedStack := &Stack{}
+	savedStack.load()
+	s.StackID = savedStack.StackID
+	s.StackName = savedStack.StackName
+	s.SSHKeyName = savedStack.SSHKeyName
+
+	go func() {
+		defer close(s.Done)
+
+		if s.promptUseExistingStack(savedStack) {
+			return
+		}
+
+		steps := []func() error{
+			s.createKeyPair,
+			s.allocateDomain,
+			s.fetchImageID,
+			s.createStack,
+			s.fetchStackOutputs,
+			s.configureDNS,
+			s.bootstrap,
+		}
+
+		for _, step := range steps {
+			if err := step(); err != nil {
+				s.persist()
+				s.HandleError(err)
+				return
+			}
+			if err := s.persist(); err != nil {
+				s.HandleError(err)
+			}
+		}
+	}()
+	return nil
+}
+
+func (s *Stack) HandleEvent(description string) {
+	s.eventMutex.Lock()
+	defer s.eventMutex.Unlock()
+	s.EventChan <- &Event{description}
+}
+
+func (s *Stack) HandleError(err error) {
+	s.errMutex.Lock()
+	defer s.errMutex.Unlock()
+	s.ErrChan <- err
+}
+
+func (s *Stack) fetchImageID() error {
+	s.HandleEvent("Fetching image manifest")
+	if err := func() error {
+		latestVersion, err := fetchLatestVersion()
+		if err != nil {
+			return err
+		}
+		var imageID string
+		for _, i := range latestVersion.Images {
+			if i.Region == s.Region {
+				imageID = i.ID
+				break
+			}
+		}
+		if imageID == "" {
+			return errors.New(fmt.Sprintf("No image found for region %s", s.Region))
+		}
+		s.ImageID = imageID
+		return nil
+	}(); err != nil {
+		s.HandleEvent(err.Error())
+		if s.ImageID != "" {
+			s.HandleEvent("Falling back to saved Image ID")
+			return nil
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Stack) allocateDomain() error {
+	s.HandleEvent("Allocating domain")
+	domain, err := AllocateDomain()
+	if err != nil {
+		return err
+	}
+	s.Domain = domain
+	return nil
+}
+
+func (s *Stack) createKeyPair() error {
+	keypairName := "flynn"
+	if s.SSHKeyName != "" {
+		keypairName = s.SSHKeyName
+	}
+	if keypair, err := loadSSHKey(keypairName); err == nil {
+		s.SSHKey = keypair
+		s.SSHKeyName = keypairName
+		res, err := s.ec2.DescribeKeyPairs(&ec2.DescribeKeyPairsRequest{
+			Filters: []ec2.Filter{{
+				Name:   aws.String("key-name"),
+				Values: []string{keypairName},
+			}},
+		})
+		if err == nil && len(res.KeyPairs) > 0 && *res.KeyPairs[0].KeyName == keypairName {
+			// key exists, we're good to go
+			// TODO(jvatic): verify key fingerprint
+			s.HandleEvent(fmt.Sprintf("Using saved key pair (%s)", keypairName))
+			return nil
+		}
+	}
+
+	s.HandleEvent("Creating key pair")
+	keypair, err := sshkeygen.Generate()
+	if err != nil {
+		return err
+	}
+
+	enc := base64.StdEncoding
+	publicKeyBytes := make([]byte, enc.EncodedLen(len(keypair.PublicKey)))
+	enc.Encode(publicKeyBytes, keypair.PublicKey)
+
+	res, err := s.ec2.ImportKeyPair(&ec2.ImportKeyPairRequest{
+		KeyName:           aws.String(keypairName),
+		PublicKeyMaterial: publicKeyBytes,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "already exists") {
```

Comment:
Also, this check should be moved up into the above if statement to reduce indentation.